File: pipeline/lib/poi_sites.py
"""Resolving co-located waypoints into SITES - a shelter with its privy, its
campsites and its water source, modelled as one place with parts (#523).

Design: ../../features/POI_SITES.md. Every number in this module was measured
against the live ATC FeatureServer, and the measurements are recorded next to
the constant they justify rather than in a commit message.

Pure module - no I/O, no DuckDB, no network. export_poi.py wires this up
against real unified records and publishes the result as three properties.

WHY THIS IS IN THE PIPELINE AND NOT THE CLIENT

A site is the thing a report, a closure or a field note references, and
lib/poi_schema.py's `unify_poi` already commits to ids that "stay stable across
repeated pipeline runs" for exactly that reason. A group computed on a phone has
no id that survives a pan, re-clusters at every zoom, and answers "how many"
when the question at a shelter is "is there a privy".

The grouping evidence is also upstream: it is ATC's own naming convention, which
lives in the raw `Name` column export_poi.py reads and the client never sees.

WHAT THE MAP DOES WITHOUT THIS

`icon-allow-overlap: false` means MapLibre never draws two colliding pins - it
drops the one that loses POI_PRIORITY, and privies lose to shelters. At zoom 14,
3% of the A.T.'s 316 privies are drawn anywhere on the trail, so the hiker sees a
clean map and concludes there is no privy. A privy sits a median 42 m from its
shelter and cannot be drawn until zoom 16; BASEMAP_MAX_ZOOM is 14. Sites remove
the pins that were colliding rather than permitting overlap, so 90% of privies
stop competing for a pin and start riding one that is actually drawn.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from lib.spurs import distance_m

# Which types can be the one anchor of a site, in the order they win it. A
# campsite anchors only what no shelter claimed - which is what makes "exactly
# one anchor per site" decidable rather than a race between two passes.
ANCHOR_TYPES = ("shelter", "campsite")

# Which types can ride an anchor's pin. `campsite` is in both tuples on purpose:
# it is a member of a shelter's site and an anchor of its own when there is no
# shelter.
MEMBER_TYPES = ("privy", "campsite", "water")

# `viewpoint`, `parking` and `resupply` are deliberately in NEITHER tuple, and
# this is the constraint that rules out grouping by distance alone. At 60 m the
# corridor holds 64 viewpoint+viewpoint clusters - two overlooks that close are
# two overlooks, not one place with parts, and a blind radius merges them.
# `parking + resupply` (26 clusters) is a trailhead: a different feature with a
# different card, and out of scope for v1 (POI_SITES.md open question 2).

# How far a name match may reach. Loose because the NAME is carrying the
# argument: every one of the name-matched privies measured is within 150 m of
# its named parent and 98% are within 100 m, so name and geometry agree
# independently on the same answer.
NAME_MATCH_RADIUS_M = 150.0

# How far proximity alone may reach, with no naming evidence at all. Tight
# because geometry is all it has. A name-only rule, by contrast, ships a
# 903 km match - a generic campsite name colliding with a same-named place at
# the far end of the trail - which is why neither gate stands alone.
PROXIMITY_RADIUS_M = 60.0

# An absolute ceiling on a site's radius, whatever gate admitted the member.
# One mile, requested on the #523 pull request.
#
# IT CANNOT BIND TODAY, AND THAT IS THE POINT OF IT. The widest gate above is
# 150 m, and the furthest member measured over the whole corridor is 143 m, so
# every real pairing clears this by a factor of ten. What it guards is the next
# edit: nothing else here stops someone raising NAME_MATCH_RADIUS_M to 2 km on a
# hunch, and the failure that reappears the moment they do is the 903 km match -
# a privy attached to a shelter at the other end of the Appalachian Trail,
# published into artifacts, with a hiker unable to undo it.
#
# So it is applied as `min()` against each gate rather than as a separate check
# further down. A guard the gates are compared against cannot be bypassed by
# widening a gate, which is the one way this particular mistake gets made.
#
# A mile is far looser than the evidence would justify as a working radius, and
# deliberately: this is a backstop, not the rule. The rule is the two gates.
MAX_SITE_RADIUS_M = 1609.344

# The trailing words ATC appends to a child's name, stripped to recover the
# parent's: "Mt. Algo Shelter Privy" -> "mt algo".
#
# EVERY ENTRY HERE IS ONE THE CORRIDOR PAYS FOR. Measured by ablation against
# all 828 shelters/campsites/privies, each word added alone on top of the first
# three:
#
#   shelters   +2 campsites, +1 site   ATC names a campsite after a PAIR:
#                                      "Tumbling Run Shelters Campsite"
#   group      +4 privies, +7 campsites, +4 sites
#                                      "Eckville Shelter Group Campsite",
#                                      "Osgood Tentsite Group Campsite"
#
# And eight that bought exactly nothing, listed so nobody adds them back on
# intuition: privies, campsites, tentsite, tentsites, campground, leanto, hut,
# cabin, site. POI_SITES.md names "Tentsite vs Campsite" as a suspected cause of
# the unresolved tail; measured, it is not one - the parent is named
# "Osgood Tentsite Group Campsite" and the child "Osgood Tentsite Privy", so
# stripping `tentsite` from BOTH sides changes nothing and stripping it from one
# would break the match.
#
# Nor is "Lean-to vs Lean to", the other suspect: punctuation is already
# collapsed to spaces before this list is consulted, so both spell "lean to" and
# folding them is dead code. It was written, measured at zero, and deleted.
TYPE_WORDS = frozenset({"privy", "campsite", "shelter", "shelters", "group"})

# Trailing digits go too: ATC numbers siblings ("Mt. Wilcox South Shelter 1",
# "Grafton Notch Parking Area Privy 2"), and 53 of the 828 names end in one.
# This is the single biggest normalisation win - privies matched go from 86% to
# 89% - and the one POI_SITES.md correctly predicted.

_NOT_ALPHANUMERIC = re.compile(r"[^a-z0-9]+")

# What `site_role` says on a published feature.
ROLE_ANCHOR = "anchor"
ROLE_MEMBER = "member"


@dataclass
class Site:
    """One place with parts: an anchor POI and the members riding its pin.

    A site with no members is not a site and `group_sites` never returns one -
    a lone shelter is just a shelter, and writing site properties onto it would
    tell the client to render a composition of one.
    """

    anchor: dict
    members: list[dict] = field(default_factory=list)

    @property
    def site_id(self) -> str:
        """The anchor's own POI id - deterministic, and already stable across
        runs because unify_poi made it so."""
        return self.anchor["id"]

    @property
    def site_name(self) -> str | None:
        return self.anchor.get("name")

    def size(self) -> int:
        return 1 + len(self.members)


def normalise_name(name: str | None) -> str:
    """Lowercase, punctuation to spaces, whitespace collapsed.

    Punctuation-to-spaces rather than punctuation-deleted, because deleting it
    would run words together: "Mt.Algo" matches nothing, while "mt algo" is what
    the shelter's own name normalises to.
    """
    if not name:
        return ""
    return " ".join(_NOT_ALPHANUMERIC.sub(" ", name.lower()).split())


def base_name(name: str | None) -> str:
    """The parent name inside a child's name: normalised, with trailing type
    words and sibling numbers removed.

    Repeated rather than single-pass, because the tokens stack - "Bald Mtn Brook
    Lean-to Privy 2" sheds a digit and then a type word to reach "bald mtn brook
    lean to", which is what the shelter's own name reduces to.
    """
    tokens = normalise_name(name).split()
    while tokens and (tokens[-1] in TYPE_WORDS or tokens[-1].isdigit()):
        tokens.pop()
    return " ".join(tokens)


def _candidate_anchors(member: dict, anchors: list[dict], by_base_name: dict[str, list[dict]]) -> list[tuple]:
    """Every anchor this member could belong to, best first.

    Sort key, in order:

    1. **Name evidence beats proximity.** A named parent is the better signal
       and the radius reflects that; a 60 m neighbour with a different name is
       the fallback, not the answer.
    2. **A contained name beats a merely reduced one.** This is the tie-break
       the corridor asked for. Stripping `group` makes "Laurel Ridge Campsite"
       and "Laurel Ridge Group Campsite" reduce to the same base, so "Laurel
       Ridge Campsite Privy" matches both - and nearest-wins picked the group
       site, 10 m closer and not what the privy is called. Preferring the anchor
       whose whole name the member's name contains picks the one ATC named it
       after.
    3. **Then nearest**, which is all geometry can offer.
    """
    member_full = normalise_name(member.get("name"))
    named: list[tuple] = []
    for candidate in by_base_name.get(base_name(member.get("name")), ()):
        if candidate["id"] == member["id"]:
            continue
        metres = distance_m(member["lat"], member["lon"], candidate["lat"], candidate["lon"])
        if metres <= min(NAME_MATCH_RADIUS_M, MAX_SITE_RADIUS_M):
            candidate_full = normalise_name(candidate.get("name"))
            contained = bool(candidate_full) and member_full.startswith(candidate_full)
            named.append((0 if contained else 1, metres, candidate))
    if named:
        return sorted(named, key=lambda entry: (entry[0], entry[1]))

    nearby: list[tuple] = []
    for candidate in anchors:
        if candidate["id"] == member["id"]:
            continue
        metres = distance_m(member["lat"], member["lon"], candidate["lat"], candidate["lon"])
        if metres <= min(PROXIMITY_RADIUS_M, MAX_SITE_RADIUS_M):
            nearby.append((1, metres, candidate))
    return sorted(nearby, key=lambda entry: (entry[0], entry[1]))


def _index_by_base_name(anchors: list[dict]) -> dict[str, list[dict]]:
    index: dict[str, list[dict]] = {}
    for anchor in anchors:
        key = base_name(anchor.get("name"))
        if key:
            index.setdefault(key, []).append(anchor)
    return index


def group_sites(records: list[dict]) -> list[Site]:
    """Fold co-located waypoints into sites.

    Two passes, shelters then the campsites no shelter claimed, which is how
    ANCHOR_TYPES' ordering becomes "exactly one anchor per site" rather than a
    contest. A member is folded once and never reconsidered: reopening a
    resolved member in the second pass would let a campsite steal a privy from
    the shelter it is named after.

    Anchors are never members of each other. Two shelters 40 m apart - Horns
    Pond has two lean-tos and is genuinely one place - stay two pins in v1,
    which is safe and slightly wrong there (POI_SITES.md open question 1).

    Returns sites in a deterministic order: by anchor id, and members by their
    own id, so repeated runs over unchanged input publish byte-identical
    properties. Stability is not cosmetic here - `verify_release.py` compares
    hashes, and a set-iteration order would rewrite every artifact for nothing.
    """
    sites: dict[str, Site] = {}
    folded: set[str] = set()

    def fold(members: list[dict], anchors: list[dict]) -> None:
        if not anchors:
            return
        by_base_name = _index_by_base_name(anchors)
        anchor_ids = {anchor["id"] for anchor in anchors}
        for member in sorted(members, key=lambda record: record["id"]):
            if member["id"] in folded or member["id"] in anchor_ids:
                continue
            candidates = _candidate_anchors(member, anchors, by_base_name)
            if not candidates:
                continue
            _, _, anchor = candidates[0]
            sites.setdefault(anchor["id"], Site(anchor=anchor)).members.append(member)
            folded.add(member["id"])

    shelters = [r for r in records if r["poi_type"] == "shelter"]
    campsites = [r for r in records if r["poi_type"] == "campsite"]
    members = [r for r in records if r["poi_type"] in MEMBER_TYPES]

    fold(members, shelters)
    # Only campsites that are not themselves members can anchor. A campsite
    # folded into a shelter's site is a part of that place, and a part cannot
    # hold parts of its own.
    fold(members, [c for c in campsites if c["id"] not in folded])

    return [sites[anchor_id] for anchor_id in sorted(sites)]


def site_properties(sites: list[Site]) -> dict[str, dict]:
    """POI id -> the three published properties, for every anchor and member.

    Additive by design (POI_SITES.md §3): a client built before this ignores
    them and behaves exactly as it does today, which is the same
    backward-compatibility rule `mile`, `capacity`, `description` and `photos`
    are already held to. A POI in no site appears nowhere in this mapping, and
    export_poi.py writes NULL for it - which is most POIs.
    """
    properties: dict[str, dict] = {}
    for site in sites:
        for record, role in [(site.anchor, ROLE_ANCHOR), *[(m, ROLE_MEMBER) for m in site.members]]:
            properties[record["id"]] = {
                "site_id": site.site_id,
                "site_role": role,
                "site_name": site.site_name,
            }
    return properties


# ---------------------------------------------------------------------------
# A SECOND GROUPING, FOR GROUND WHERE NOBODY NAMED THE PARTS (#1536)
#
# Everything above resolves ATC's naming convention: "Mt. Algo Shelter Privy"
# names its parent, so the NAME carries the argument and PROXIMITY_RADIUS_M is
# only the sanity check on it. New York City's fountains have no such
# convention and no parent to name - 3,148 of them carry a park name, 912
# distinct, and Central Park's 137 all carry the same one. A park name says
# "same property", not "same place".
#
# So the two gates swap weights. Geometry carries the argument here and the
# name is the sanity check, which is why this is a second function rather than
# a looser radius on the first: widening NAME_MATCH_RADIUS_M would loosen
# ATC's grouping too, and the 903 km match its docstring warns about is exactly
# what a name-led rule ships when the name stops being specific.


#: How far apart two waypoints on one property may be and still be one place.
#:
#: MEASURED against the fountains this exists for, 2026-09-17, release
#: 2026-09-16-4: over the 2,823 nearest-neighbour distances between fountains
#: sharing a park property, p25 is 24 m, the MEDIAN IS 41 m, p75 is 73 m and
#: p90 is 124 m. So 80 m sits just above the third quartile - it folds the
#: banks of fountains a playground or a plaza holds, and leaves the quarter of
#: pairings that are genuinely spread across a park.
#:
#: Wider was available and refused. A 150 m gate admits 93% of pairings rather
#: than 78%, and 150 m is what NAME_MATCH_RADIUS_M already allows above - but
#: it allows it on ATC's evidence, where the name names one parent. Here the
#: same 150 m would start folding opposite ends of a playground into one mark
#: on the strength of both being in the same park, which is not evidence that
#: they are one place.
#:
#: The failure this bounds is a hiker walking to a pin and not finding the
#: fountain. That is bounded by the anchor rule rather than by this number -
#: see `group_place_sites`, which draws the pin on a REAL member and never on
#: a centroid - but a tighter gate keeps the walk from the pin to the next
#: member short as well.
PLACE_PROXIMITY_RADIUS_M = 80.0


def group_place_sites(
    records: list[dict],
    radius_m: float = PLACE_PROXIMITY_RADIUS_M,
    types: tuple[str, ...] = ("water", "privy"),
) -> list[Site]:
    """Fold waypoints of ONE type that share a place name and stand together.

    Single-link within each (name, type) group, so a line of fountains down a
    boardwalk folds as one site instead of a chain of overlapping pairs - the
    Coney Island case, where no two ends are within the radius of each other
    but every step is.

    THE ANCHOR IS A REAL WAYPOINT, NEVER A CENTROID, and that is the safety
    rule of this whole function. A hiker walks to the pin. A centroid of twelve
    fountains is a point on a lawn with no fountain on it, which is the
    confidently wrong answer FEATURES.md calls more dangerous than an honest
    unknown. The anchor is the member CLOSEST to the group's centre - the most
    central fountain that actually exists - so the pin is always somewhere a
    hiker can arrive and find water.

    Ties break on the id, so the same input gives the same site id on every
    run - which is what a report or a field note pointing at a site needs.
    """
    wanted = set(types)
    groups: dict[tuple[str, str], list[dict]] = {}
    for record in records:
        if record.get("poi_type") not in wanted:
            continue
        name = record.get("name")
        if not name:
            # No name is no gate. Geometry alone is what PROXIMITY_RADIUS_M
            # above exists to refuse - two overlooks that close are two
            # overlooks - and an unnamed fountain has nothing to pair that
            # geometry with, so it keeps its own pin.
            continue
        groups.setdefault((normalise_name(name), record["poi_type"]), []).append(record)

    sites: list[Site] = []
    for members in groups.values():
        if len(members) < 2:
            continue
        for cluster in _single_link(members, radius_m):
            if len(cluster) < 2:
                continue
            anchor = _most_central(cluster)
            sites.append(Site(anchor=anchor, members=[m for m in cluster if m is not anchor]))
    return sites


def _single_link(records: list[dict], radius_m: float) -> list[list[dict]]:
    """Union-find over the pairs within `radius_m`, returning each component."""
    parent = list(range(len(records)))

    def find(i: int) -> int:
        while parent[i] != i:
            parent[i] = parent[parent[i]]
            i = parent[i]
        return i

    for i in range(len(records)):
        for j in range(i + 1, len(records)):
            here, there = records[i], records[j]
            if distance_m(here["lat"], here["lon"], there["lat"], there["lon"]) <= radius_m:
                a, b = find(i), find(j)
                if a != b:
                    parent[a] = b

    clusters: dict[int, list[dict]] = {}
    for index, record in enumerate(records):
        clusters.setdefault(find(index), []).append(record)
    return list(clusters.values())


def _most_central(cluster: list[dict]) -> dict:
    """The member nearest the cluster's mean position, id breaking ties."""
    mid_lat = sum(r["lat"] for r in cluster) / len(cluster)
    mid_lon = sum(r["lon"] for r in cluster) / len(cluster)
    return min(cluster, key=lambda r: (distance_m(r["lat"], r["lon"], mid_lat, mid_lon), str(r["id"])))

File: client/src/map/poiSites.ts
// One pin per SITE rather than one per POI (#524, features/POI_SITES.md).
//
// THE PROBLEM IS NOT OVERLAP, IT IS DELETION. map/poiLayers.ts sets
// `icon-allow-overlap: false` deliberately, so MapLibre draws no two colliding
// pins - it drops whichever loses POI_PRIORITY. Privies lose to shelters and sit
// a median 42 m from theirs, so simulating placement over the whole corridor puts
// the share of privies drawn ANYWHERE on the trail at 3% at zoom 14, and
// `BASEMAP_MAX_ZOOM` is 14, so that is the zoom an offline hiker lives at. The
// privy layer, added at real download cost in #510, effectively does not exist.
// A hiker sees a clean map and concludes there is no privy.
//
// That is not a bug in POI_PRIORITY. It is the right answer to the wrong
// question: where two pins describe PARTS OF THE SAME PLACE, "which one
// survives" has no acceptable answer. So the members stop competing. They are
// removed from the source, and the anchor's pin says what is there.
//
// The measured effect, from #523's grouping over the live corridor: 284 of 316
// privies and 144 of 232 campsites stop asking for a box, and start riding a pin
// that places at 91% at z14.
//
// WHAT THIS DELIBERATELY IS NOT:
//
//  - Not a change to `icon-allow-overlap`. Letting pins overlap is the symptom
//    the original complaint started from. Sites remove the colliding pins; the
//    collision engine keeps doing its job on what is left, and two shelters
//    400 m apart still collide at z12, correctly.
//  - Not a clustering layer. A crowded ridge of viewpoints is still the
//    collision engine's problem, and #531's is the zoom tiering.
//  - Not spiderfying. Fanning members out on leader lines draws every member at
//    a position it is not at. This app refuses to draw a stale GPS fix like a
//    live one; drawing a privy 80 px from where it is, is the same refusal.

import type { MapPoint } from '../lib/legendContents'

/** What `site_role` says on a published feature - pipeline/lib/poi_sites.py. */
export const SITE_ROLE_ANCHOR = 'anchor'
export const SITE_ROLE_MEMBER = 'member'

/**
 * The feature property naming what a site pin carries.
 *
 * A property rather than a separate layer, for the reason POI_ID_PROPERTY is one:
 * the style has to reach it from a `match` expression, and MapLibre gives an
 * expression the properties of the feature it is placing and nothing else.
 */
export const SITE_MEMBERS_PROPERTY = 'site_members'

/**
 * The categories that can ride an anchor's pin, in the order a pin lists them.
 *
 * The same three `MEMBER_TYPES` the pipeline groups on, and the order is fixed
 * here rather than taken from the data so that one site always produces one icon
 * id. Sorted by what a hiker asks first at a shelter: features/POI_SITES.md's own
 * framing is "is there a privy, and is there water", and a campsite is the one of
 * the three you can see for yourself once you are standing there.
 */
export const SITE_MEMBER_TYPES = ['privy', 'water', 'campsite'] as const

/**
 * The categories that can ANCHOR a site - pipeline/lib/poi_sites.py's
 * ANCHOR_TYPES.
 *
 * Only these get site pin variants built, because only these can ever carry a
 * footer strip. A viewpoint given the full member matrix would be fourteen
 * images the style can never ask for.
 */
export const SITE_ANCHOR_TYPES = ['shelter', 'campsite'] as const

/**
 * A map point that may belong to a site.
 *
 * All three fields optional, and for the reason `StoredPoi.source` is: a phone
 * that downloaded before #523 published them has POIs without any, and undefined
 * means "this copy predates the grouping" rather than "this POI is not in a
 * site". Both come out as a plain pin, which is exactly what that phone drew
 * before, so an old download degrades to the old behaviour instead of to a blank
 * map.
 */
export interface SitePoint extends MapPoint {
  siteId?: string
  siteRole?: string
}

export interface SiteComposition {
  /** The points that get a pin of their own: every anchor, and everything that
   *  is not in a site at all. */
  drawn: readonly SitePoint[]
  /** Per anchor id, the distinct member categories riding its pin, in
   *  `SITE_MEMBER_TYPES` order. Absent for a point that carries nothing. */
  membersFor: ReadonlyMap<string, readonly string[]>
}

/**
 * Which points draw a pin, and what each site pin has to say.
 *
 * A MEMBER WHOSE ANCHOR IS NOT HERE KEEPS ITS OWN PIN. That is the case worth
 * spelling out: dropping a member on the strength of its `site_id` alone would
 * mean that anything removing the anchor from this list - a future filter, a
 * partial download, a grouping written against a POI this build never received -
 * silently removes the privy too. The hiker would be left with neither pin and no
 * way to discover either. So the drop is conditional on the pin that replaces it
 * actually being drawn.
 *
 * Members are counted by DISTINCT CATEGORY, not by count. A site with four
 * campsites says "campsite", not "campsite ×4": the question at a shelter is
 * whether there is one, and a number answers a question nobody asked - the same
 * reasoning features/POI_SITES.md gives for rejecting a bare `+N` bubble.
 */
export function composeSites(points: readonly SitePoint[]): SiteComposition {
  const anchors = new Set<string>()
  for (const point of points) {
    if (point.siteRole === SITE_ROLE_ANCHOR && point.siteId !== undefined) {
      anchors.add(point.siteId)
    }
  }

  const members = new Map<string, Set<string>>()
  const drawn: SitePoint[] = []

  for (const point of points) {
    const ridesAnchor =
      point.siteRole === SITE_ROLE_MEMBER &&
      point.siteId !== undefined &&
      anchors.has(point.siteId)

    if (!ridesAnchor) {
      drawn.push(point)
      continue
    }

    const carried = members.get(point.siteId!) ?? new Set<string>()
    carried.add(point.type)
    members.set(point.siteId!, carried)
  }

  // Keyed by ANCHOR ID rather than site id, because that is what the feature
  // collection has to hand when it writes the property - one lookup at build
  // time instead of a second index from site id back to anchor.
  const membersFor = new Map<string, readonly string[]>()
  for (const point of drawn) {
    if (point.siteRole !== SITE_ROLE_ANCHOR || point.siteId === undefined) continue
    const carried = members.get(point.siteId)
    if (carried === undefined || carried.size === 0) continue
    membersFor.set(
      point.id,
      SITE_MEMBER_TYPES.filter((type) => carried.has(type)),
    )
  }

  return { drawn, membersFor }
}

/**
 * Every POI at one site, the anchor first, for the card that has to lead to
 * them (#526).
 *
 * A SIBLING OF `composeSites` RATHER THAN A SECOND READING OF IT. That function
 * answers what a PIN says and deliberately throws away everything a card needs:
 * its members are distinct category strings keyed by anchor id, so a site with
 * two privies collapses to one entry and neither privy has an id left to open.
 * Correct for a 38 px pin, fatal for a strip that has to lead somewhere - the
 * "Backpacker Campsite Upper Privy" / "...Lower Privy" pair in
 * features/POI_SITES.md's open question 4 is two real points at one place, and
 * both must be reachable. So this keys on POI id and returns the points
 * themselves.
 *
 * Generic over the point type, which is what lets the shell pass its stored
 * POIs - names, photos, descriptions and all - and get them back without an
 * adapter type in between, while this file still knows nothing about
 * lib/trailData.ts.
 *
 * IT MIRRORS `composeSites`' RULE, IN REVERSE. There is no site here unless the
 * anchor is: a member whose anchor is missing from `points` - a partial
 * download, a future filter - is drawing its own pin, so listing it as part of
 * something the map is not drawing would be a strip hanging off nothing. Same
 * for a role this build does not recognise: the point keeps its own pin over
 * there, so its card is the plain card, not a site's.
 */
export function siteRoster<T extends SitePoint>(
  points: readonly T[],
  poiId: string,
): readonly T[] {
  const found = points.find((point) => point.id === poiId)
  if (found === undefined || found.siteId === undefined) return []
  if (found.siteRole !== SITE_ROLE_ANCHOR && found.siteRole !== SITE_ROLE_MEMBER)
    return []

  const siteId = found.siteId
  const anchor = points.find(
    (point) => point.siteId === siteId && point.siteRole === SITE_ROLE_ANCHOR,
  )
  if (anchor === undefined) return []

  const members = points.filter(
    (point) => point.siteId === siteId && point.siteRole === SITE_ROLE_MEMBER,
  )
  members.sort(
    (a, b) => memberRank(a.type) - memberRank(b.type) || compareIds(a.id, b.id),
  )

  return [anchor, ...members]
}

/**
 * Where a member type sorts in the strip - `SITE_MEMBER_TYPES` order, and
 * anything else after it rather than dropped.
 *
 * Filtering to the known three would be the shorter line and would make a
 * member type a later release publishes unreachable from its own site's card,
 * which is the exact bug #526 exists to fix. An unfamiliar category sorts last
 * and still gets a chip; the neutral diamond pin MapIcon draws for it says it
 * is something this build has not heard of.
 */
function memberRank(type: string): number {
  const at = (SITE_MEMBER_TYPES as readonly string[]).indexOf(type)
  return at === -1 ? SITE_MEMBER_TYPES.length : at
}

/** A stable tiebreak inside one category, so two privies at one campsite come
 *  out in the same order on every render. By id and not by name because a
 *  `SitePoint` carries no name - this file's whole input is geometry and site
 *  keys - and an unstable strip is worse than an arbitrary one. */
function compareIds(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0
}

/** Metres per degree of latitude, and of longitude at the equator -
 *  pipeline/lib/spurs.py's `METERS_PER_DEGREE`. */
const METERS_PER_DEGREE = 111_320

/**
 * How far apart two points at one site are, in metres.
 *
 * Equirectangular, not haversine, and copied from the pipeline's `distance_m`
 * (pipeline/lib/spurs.py) on purpose: THAT is the measurement that decided this
 * member belongs to this site - a 60 m proximity gate and a 150 m name gate,
 * with the furthest real member at 143 m. A second formula here would put a
 * different number on the card from the one that admitted the point, which is
 * the drift map/MapIcon.ts's header calls the one failure a symbol cannot
 * survive, in arithmetic. Its own docstring gives the accuracy: under a
 * kilometre the flat approximation is good to well under 1%, and `cos()` at the
 * midpoint latitude is what keeps a longitude degree honest across the trail's
 * 34-46 degree span.
 */
export function siteDistanceMeters(
  from: { lat: number; lon: number },
  to: { lat: number; lon: number },
): number {
  const meanLat = (((from.lat + to.lat) / 2) * Math.PI) / 180
  const dy = (to.lat - from.lat) * METERS_PER_DEGREE
  const dx = (to.lon - from.lon) * METERS_PER_DEGREE * Math.cos(meanLat)
  return Math.hypot(dx, dy)
}

/**
 * The member list as one property value, for the style to `match` on.
 *
 * A sorted, joined string rather than an array because a MapLibre `match`
 * expression compares scalars, and the icon matrix is keyed on exactly this -
 * see map/poiIcons.ts. Empty string for a pin carrying nothing, so the property
 * is always present and the expression needs no `coalesce`.
 */
export function siteMembersKey(members: readonly string[] | undefined): string {
  return members === undefined ? '' : members.join('+')
}
